### ISSUE TITLE
Commit found, but unreachable

### ISSUE THREAD

**Issue by RalfJung** [ocaml/opam#3548]
After upgrading to opam 2, our CI fails with
```
[ERROR] Could not synchronize /builds/FP/iris-coq/opamroot/default/.opam-switch/sources/coq-mtac from "git+https://github.com/Mtac2/Mtac2#9c9f628b87a639989fbec473e3508ac8e05f5f62":
        Commit found, but unreachable: enable uploadpack.allowReachableSHA1InWant on server
```
This happens when pinning `https://github.com/Mtac2/Mtac2#9c9f628b87a639989fbec473e3508ac8e05f5f62`. The same pin worked fine with opam 1.

The strange thing is that this commit is *not* unreachable. It is about 20 commits back from [`master-8.8`](https://github.com/Mtac2/Mtac2/commits/master-8.8).

**Comment by RalfJung** [comment]
Interestingly, this works fine on my local system. My laptop has Debian testing (git version 2.18.0), the CI uses Debian stable (git version 2.11.0). Might be that opam is expecting too much from this old version of git?

**Comment by rjbou** [comment]
For git older to 2.14, there is a problem on reaching single commits. You can find explanation in this issue #3523.
Thanks for reporting.

**Comment by RalfJung** [comment]
Indeed I can confirm that after updating CI to git 2.18, this works now.

Closing as duplicate of https://github.com/ocaml/opam/issues/3523.

**Comment by RalfJung** [comment]
Actually I have to take this back, we are seeing the issue again with git 2.18 now.

I even just had it on my laptop once, where what I did was:
* I already had `coq.8.8.2` installed.
* I did `opam pin add coq-mtac.dev -k git https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19` but then hit Ctrl-C.
* I did `opam pin -vv add coq-mtac.dev -k git https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19` and it showed
```
+ /usr/bin/ocamlc "-vnum"
- 4.05.0
+ /usr/bin/git "remote" "set-url" "origin" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" "--update-shallow" "+ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
- error: Server does not allow request for unadvertised object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" "+ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
- error: Server does not allow request for unadvertised object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "update-ref" "refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" "ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
- fatal: update_ref failed for ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19': cannot update ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19': trying to write ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19' with nonexistent object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "fetch" "-q" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "show" "-s" "--format=%H" "ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac)
- ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
[ERROR] Could not synchronize /home/r/.opam/coq-8.8/.opam-switch/sources/coq-mtac from "git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19":
        Commit found, but unreachable: enable uploadpack.allowReachableSHA1InWant on server
[ERROR] Error getting source from git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:
          - git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
```
However so far I was unable to reproduce the failure on my laptop, it happened just once... I assume it has something to do with the state of the cache?

**Comment by rjbou** [comment]
At this level, it's git cache, not opam cache. In the log, there is a git error `error: Server does not allow request for unadvertised object [...]`, which leads to the opam error (this time explaining the good state).

**Comment by RalfJung** [comment]
But why does `opam pin` then work locally?

And indeed if I try these commands separately, git says "server does not allow", but if I run `opam pin -vv` now it actually claims that this command would succeed! Does opam somehow feed existing cached stuff into the new repo it is creating? How did it manage to ever download this commit then, though?

**Comment by RalfJung** [comment]
I can actually reproduce this with an empty `OPAMROOT`:
`opam pin add -vv -n coq-mtac.dev -k git https://github.com/Mtac2` first fails:
```
/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/ocamlc "-vnum"
- 4.05.0
Package coq-mtac does not exist, create as a NEW package? [Y/n] y
+ /usr/bin/git "init" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- Initialized empty Git repository in /home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac/.git/
+ /usr/bin/git "config" "--local" "fetch.prune" "false" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "remote" "add" "origin" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "init" "--bare" (CWD=/home/r/Desktop/tmo/download-cache/git)
- Initialized empty Git repository in /home/r/Desktop/tmo/download-cache/git/
+ /usr/bin/git "remote" "set-url" "origin" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" "--update-shallow" "+ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- error: Server does not allow request for unadvertised object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" "+ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- error: Server does not allow request for unadvertised object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "update-ref" "refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" "ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- fatal: update_ref failed for ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19': cannot update ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19': trying to write ref 'refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19' with nonexistent object ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/git "fetch" "-q" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "show" "-s" "--format=%H" "ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
[ERROR] Could not synchronize /home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac from "git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19":
        Commit found, but unreachable: enable uploadpack.allowReachableSHA1InWant on server
[ERROR] Error getting source from git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:
          - git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
```
But then if I try exactly the same command again it succeeds saying
```
/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
+ /usr/bin/ocamlc "-vnum"
- 4.05.0
Package coq-mtac does not exist, create as a NEW package? [Y/n] y
+ /usr/bin/git "remote" "set-url" "origin" "https://github.com/Mtac2/Mtac2" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "fetch" "-q" "https://github.com/Mtac2/Mtac2" "--update-shallow" "+ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "push" "/home/r/Desktop/tmo/download-cache/git" "+refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19:refs/remotes/0e806cdf06f358e8c93ddb18d4aaef25" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- To /home/r/Desktop/tmo/download-cache/git
-  * [new branch]        opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19 -> 0e806cdf06f358e8c93ddb18d4aaef25
+ /usr/bin/git "diff" "--no-ext-diff" "--quiet" "refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" "--" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
+ /usr/bin/git "reset" "--hard" "refs/remotes/opam-ref-ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19" "--" (CWD=/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac)
- HEAD is now at ef46cb8b Introduce `TT.with_goal_[prop|type]`.
[coq-mtac.dev] synchronised from git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19
coq-mtac is now pinned to git+https://github.com/Mtac2/Mtac2#ef46cb8b5f7ff013df20996f4fd3cbea2d0fbc19 (version dev)
```

**Comment by RalfJung** [comment]
For now I am working around this by just pinning twice on our CI, that seems to work. Needless to say that's a bad hack though^^

**Comment by rjbou** [comment]
This hack is not to far from the fix :D 
What happened: Broadly speaking, opam uses git this way, try to fetch the specific commit/branch head,  a fallback mechanism if unreachable, and if still unreachable checking if the commit exists to display the good error  message. In order to check it, a complete fetch of the repo is done (retrieve all refs instead of `git fetch <origin_url>`which get only `FETCH_HEAD`). As the first pin fail, a complete fetch was done and as the source directory  is not removed (only temporary directories/files are, source dir here is `/home/r/Desktop/tmo/default/.opam-switch/sources/coq-mtac`), on the second pin git is able to fetch directly the specific commit (all refs are advertised from previous failure).
The fix for git < 2.14 is to retrieve all refs as a fallback mechanism, instead of just `FETCH_HEAD`, and that's what is doing the double-pinning.

**Comment by RalfJung** [comment]
Ah... yeah I already wondered why the `git show` worked. Might be worth doing the complete fetch not just as part of giving up? ;)

**Comment by RalfJung** [comment]
Awesome. :D Is there an ETA for a release?

**Comment by rjbou** [comment]
Next week :)

**Comment by sgrove** [comment]
We're hitting this with opam 2 on our CI servers as well, git version 2.11.0, and also see that if we try to pin twice it works. Looking forward to a more robust release! 😄

**Comment by hannesm** [comment]
@sgrove AFAIU solved by https://github.com/ocaml/opam/pull/3561 which is part of opam 2.0.1 https://github.com/ocaml/opam/releases/tag/2.0.1 - are you using this on your CI machines already?

**Comment by samoht** [comment]
Once the binary is ready, just need to bump https://github.com/ocaml/ocaml-ci-scripts/blob/master/.travis-ocaml.sh#L64 and https://github.com/ocaml/ocaml-ci-scripts/blob/master/.travis-ocaml.sh#L67
